lib/internal/modules/esm/import_map.js

```diff
@@ -0,0 +1,107 @@
+'use strict';
+const { isURL, URL } = require('internal/url');
+const { ObjectEntries, ObjectKeys, SafeMap, ArrayIsArray } = primordials;
+const { codes: { ERR_INVALID_IMPORT_MAP } } = require('internal/errors');
+
+class ImportMap {
+  #baseURL;
+  imports = new SafeMap();
+  scopes = new SafeMap();
+
+  constructor(raw, baseURL) {
+    this.#baseURL = baseURL;
+    processImportMap(this, this.#baseURL, raw);
+  }
+
+  get baseURL() {
+    return this.#baseURL;
+  }
+
+  resolve(specifier, parentURL = this.baseURL) {
+    // Process scopes
+    for (const { 0: prefix, 1: mapping } of this.scopes) {
+      let mappedSpecifier = mapping.get(specifier);
+      if (parentURL.pathname.startsWith(prefix.pathname) && mappedSpecifier) {
+        if (!isURL(mappedSpecifier)) {
+          mappedSpecifier = new URL(mappedSpecifier, this.baseURL);
+          mapping.set(specifier, mappedSpecifier);
+        }
+        specifier = mappedSpecifier;
+        break;
+      }
+    }
+
+    let spec = specifier;
+    if (isURL(specifier)) {
+      spec = specifier.pathname;
+    }
```

Comment:
```suggestion
    const spec = isURL(specifier) ? specifier.pathname : specifier;
```

Reply:
```suggestion
    const spec = specifier?.pathname || specifier;
```

Reply:
could it be a non-URL object?

Reply:
@JakobJingleheimer I find @ljharb's version much more understandable. Are you suggesting this form to avoid the `isURL` call? Or some other reason. Just wanted to understand before deciding on a direction. My main concern with this duck typing approach is that it is less explicit than a real check, but if there is a good reason I am happy to change it.

Reply:
Going to mark this resolved for now. I think it is satisfactory to use the form with the explicit check. We can revisit if necessary later.